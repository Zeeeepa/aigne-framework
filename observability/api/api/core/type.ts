import type { InferInsertModel } from "drizzle-orm";
import { z } from "zod";
import type { Trace } from "../server/models/trace.js";

export const AIGNEObserverOptionsSchema = z
  .object({ storage: z.string().optional() })
  .optional()
  .default({});

export type AIGNEObserverOptions = z.infer<typeof AIGNEObserverOptionsSchema>;

export type TraceFormatSpans = Omit<InferInsertModel<typeof Trace>, "id" | "rootId"> & {
  id: string;
  rootId: string;
};
