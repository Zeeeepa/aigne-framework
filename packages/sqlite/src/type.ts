import { customType } from "drizzle-orm/sqlite-core";

export const json = customType<{
  data: unknown;
  driverData: string;
}>({
  dataType: () => "json",
  fromDriver: (value) => JSON.parse(value),
  toDriver: (value) => JSON.stringify(value),
});

export const datetime = customType<{
  data: Date;
  driverData: string;
}>({
  dataType: () => "datetime",
  fromDriver: (value) => new Date(value),
  toDriver: (value) => value.toISOString(),
});
