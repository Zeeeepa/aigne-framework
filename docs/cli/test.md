# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
