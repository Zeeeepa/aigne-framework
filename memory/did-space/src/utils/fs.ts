export async function streamToString(
  stream: ReadableStream | NodeJS.ReadableStream,
  encoding: BufferEncoding = "utf-8",
): Promise<string> {
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString(encoding);
}
