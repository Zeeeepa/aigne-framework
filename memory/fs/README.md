# @aigne/fs-memory

<p align="center">
  <picture>
    <source srcset="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo-dark.svg" media="(prefers-color-scheme: dark)">
    <source srcset="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo.svg" media="(prefers-color-scheme: light)">
    <img src="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo.svg" alt="AIGNE Logo" width="400" />
  </picture>
</p>

[![GitHub star chart](https://img.shields.io/github/stars/AIGNE-io/aigne-framework?style=flat-square)](https://star-history.com/#AIGNE-io/aigne-framework)
[![Open Issues](https://img.shields.io/github/issues-raw/AIGNE-io/aigne-framework?style=flat-square)](https://github.com/AIGNE-io/aigne-framework/issues)
[![codecov](https://codecov.io/gh/AIGNE-io/aigne-framework/graph/badge.svg?token=DO07834RQL)](https://codecov.io/gh/AIGNE-io/aigne-framework)
[![NPM Version](https://img.shields.io/npm/v/@aigne/fs-memory)](https://www.npmjs.com/package/@aigne/fs-memory)
[![Elastic-2.0 licensed](https://img.shields.io/npm/l/@aigne/fs-memory)](https://github.com/AIGNE-io/aigne-framework/blob/main/LICENSE)

File system memory component for [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing file system-based memory storage capabilities.

## Introduction

`@aigne/fs-memory` is the file system memory component of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing local file system-based memory storage and retrieval functionality. This component stores memories as YAML files in specified directories, providing a simple and reliable memory persistence solution.

## Features

* **File System Storage**: Uses local file system for memory file storage
* **YAML Format**: Uses YAML format for memory storage, making it easy to read and edit
* **Directory Management**: Automatically creates and manages memory directory structure
* **Path Support**: Supports absolute paths, relative paths, and home directory (~) expansion
* **AI Agent Management**: Uses AI agents to handle memory recording and retrieval
* **TypeScript Support**: Complete type definitions providing an excellent development experience

## Installation

### Using npm

```bash
npm install @aigne/fs-memory
```

### Using yarn

```bash
yarn add @aigne/fs-memory
```

### Using pnpm

```bash
pnpm add @aigne/fs-memory
```

## Basic Usage

```typescript
import { AIAgent, AIGNE } from "@aigne/core";
import { FSMemory } from "@aigne/fs-memory";
import { OpenAIChatModel } from "@aigne/openai";

// Create AI model instance
const model = new OpenAIChatModel({
  apiKey: process.env.OPENAI_API_KEY,
  model: "gpt-4-turbo",
});

// Create file system memory
const memory = new FSMemory({
  rootDir: "./memories",
});

// Create AI agent with file-based memory
const agent = AIAgent.from({
  name: "FileAssistant",
  instructions: "You are a helpful assistant with file-based memory.",
  memory: memory,
  inputKey: "message",
});

// Use AIGNE execution engine
const aigne = new AIGNE({ model });

// Invoke agent
const userAgent = await aigne.invoke(agent);

// Send message
const response = await userAgent.invoke({
  message: "Remember that I prefer working in the morning",
});

console.log(response.message);

// Query memory
const response2 = await userAgent.invoke({
  message: "What do you know about my work preferences?",
});

console.log(response2.message);
```

## Advanced Configuration

### Custom Memory Directory

```typescript
import { join } from "node:path";
import { FSMemory } from "@aigne/fs-memory";

// Using absolute path
const memory1 = new FSMemory({
  rootDir: "/absolute/path/to/memories",
});

// Using relative path
const memory2 = new FSMemory({
  rootDir: "./data/memories",
});

// Using home directory
const memory3 = new FSMemory({
  rootDir: "~/my-app/memories",
});

// Using dynamic path
const memory4 = new FSMemory({
  rootDir: join(process.cwd(), "memories"),
});
```

### Configure Memory Agent Options

```typescript
const memory = new FSMemory({
  rootDir: "./memories",
  // Custom memory retriever agent configuration
  retrieverOptions: {
    name: "CustomRetriever",
    instructions: "Custom retrieval instructions for file-based memory",
  },
  // Custom memory recorder agent configuration
  recorderOptions: {
    name: "CustomRecorder",
    instructions: "Custom recording instructions for file-based memory",
  },
  // Other agent configurations
  autoUpdate: true,
});
```

## Using Multiple Memory Types Together

```typescript
import { DefaultMemory } from "@aigne/default-memory";
import { FSMemory } from "@aigne/fs-memory";

// Combine multiple memory types
const agent = AIAgent.from({
  name: "MultiMemoryAgent",
  instructions: "You are an assistant with multiple memory types.",
  memory: [
    // Database memory for structured data
    new DefaultMemory({
      storage: {
        url: "file:memory.db",
      },
    }),
    // File system memory for long-term storage
    new FSMemory({
      rootDir: "./long-term-memories",
    }),
  ],
  inputKey: "message",
});
```

## Memory File Structure

FSMemory creates the following structure in the specified directory:

```
memories/
├── conversation-id-1/
│   └── memory.yaml
├── conversation-id-2/
│   └── memory.yaml
└── ...
```

Each memory file contains YAML-formatted memory data:

```yaml
- id: memory-id-1
  content: "User prefers working in the morning"
  timestamp: 2024-01-01T10:00:00Z
  metadata:
    type: "preference"

- id: memory-id-2
  content: "User is a software developer"
  timestamp: 2024-01-01T10:05:00Z
  metadata:
    type: "personal_info"
```

## License

Elastic-2.0
