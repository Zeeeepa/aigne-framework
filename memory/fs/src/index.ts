export * from "./memory.js";
export * from "./utils/index.js";
