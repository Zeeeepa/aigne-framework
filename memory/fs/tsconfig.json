{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {
      "@aigne/fs-memory": ["./src"]
    }
  },
  "exclude": ["node_modules", "lib"],
  "include": ["src", "test"]
}
