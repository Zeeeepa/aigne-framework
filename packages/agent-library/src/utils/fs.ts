import { stat } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";

export async function exists(path: string): Promise<boolean> {
  try {
    await stat(path);
    return true;
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      return false;
    }
    throw e;
  }
}

export function expandHome(filepath: string): string {
  if (filepath.startsWith("~/") || filepath === "~") {
    return join(homedir(), filepath.slice(1));
  }
  return filepath;
}
