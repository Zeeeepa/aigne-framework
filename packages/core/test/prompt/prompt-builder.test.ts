import { expect, test } from "bun:test";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import {
  AIAgent,
  AIAgentToolChoice,
  AIGNE,
  FunctionAgent,
  MCPAgent,
  PromptBuilder,
  TeamAgent,
} from "@aigne/core";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import type { GetPromptResult } from "@modelcontextprotocol/sdk/types.js";
import { z } from "zod";
import { MockMemory } from "../_mocks/mock-memory.js";

test("PromptBuilder should build messages correctly", async () => {
  const context = new AIGNE().newContext();

  const builder = PromptBuilder.from("Test instructions");

  const memory = new MockMemory({});
  await memory.record(
    {
      role: "agent",
      content: [{ message: "Hello, How can I help you?" }],
      source: "TestAgent",
    },

    context,
  );

  const agent = AIAgent.from({
    memory,
    inputKey: "message",
  });

  const prompt1 = await builder.build({
    agent,
    input: { message: "Hello" },
    context,
  });

  expect(prompt1.messages).toEqual([
    {
      role: "system",
      content: "Test instructions",
    },
    {
      role: "system",
      content: expect.stringContaining("Hello, How can I help you?"),
    },
    {
      role: "user",
      content: "Hello",
    },
  ]);

  const prompt2 = await builder.build({
    input: { name: "foo" },
    context,
  });
  expect(prompt2.messages).toEqual([
    {
      role: "system",
      content: "Test instructions",
    },
  ]);
});

test("PromptBuilder should build response format correctly", async () => {
  const context = new AIGNE().newContext();

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    outputSchema: z.object({
      name: z.string(),
      age: z.number().optional(),
    }),
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.responseFormat).toEqual({
    type: "json_schema",
    jsonSchema: {
      name: "output",
      schema: expect.objectContaining({
        type: "object",
        properties: {
          name: { type: "string" },
          age: { type: "number" },
        },
        required: ["name"],
        additionalProperties: false,
      }),
      strict: true,
    },
  });
});

test("PromptBuilder should build skills correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
    inputSchema: z.object({
      name: z.string(),
      age: z.number().optional(),
    }),
  });

  const teamAgentSkill = TeamAgent.from({
    name: "TestTeamAgent",
    skills: [
      AIAgent.from({
        name: "TestSkillInTeamAgent1",
      }),
      AIAgent.from({
        name: "TestSkillInTeamAgent2",
      }),
    ],
  });

  // MCPAgent is not invokable, so we use it as a skill container
  const mcpAgentSkill = new MCPAgent({
    name: "TestMCPAgent",
    client: new Client({
      name: "TestClient",
      version: "1.0.0",
    }),
    skills: [
      AIAgent.from({
        name: "TestMCPSkill1",
        instructions: "TestMCPSkill1 instructions",
      }),
      AIAgent.from({
        name: "TestMCPSkill2",
        instructions: "TestMCPSkill2 instructions",
      }),
    ],
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill, teamAgentSkill, mcpAgentSkill],
    toolChoice: skill,
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.tools).toEqual([
    {
      type: "function",
      function: {
        name: "TestSkill",
        description: "Test skill description",
        parameters: expect.objectContaining({
          type: "object",
          properties: {
            name: { type: "string" },
            age: { type: "number" },
          },
          required: ["name"],
          additionalProperties: false,
        }),
      },
    },
    // TeamAgent as skill, nested skills should no longer be included in the tools
    {
      type: "function",
      function: {
        name: "TestTeamAgent",
        parameters: {},
      },
    },
    // MCPAgent is not invokable, so it should not be included in the tools, and the nested skills should be included
    {
      type: "function",
      function: {
        name: "TestMCPSkill1",
        parameters: {},
      },
    },
    {
      type: "function",
      function: {
        name: "TestMCPSkill2",
        parameters: {},
      },
    },
  ]);

  expect(prompt.toolChoice).toEqual({
    type: "function",
    function: {
      name: skill.name,
      description: skill.description,
    },
  });
});

test("PromptBuilder should unique skills correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill, skill],
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.tools).toHaveLength(1);
});

test("PromptBuilder should build toolChoice with router mode correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill],
    toolChoice: AIAgentToolChoice.router,
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.toolChoice).toEqual("required");
});

test("PromptBuilder from string", async () => {
  const context = new AIGNE().newContext();

  const builder = PromptBuilder.from("Hello, {{agentName}}!");

  const prompt = await builder.build({ input: { agentName: "Alice" }, context });

  expect(prompt).toEqual({
    messages: [
      {
        role: "system",
        content: "Hello, Alice!",
      },
    ],
  });
});

test("PromptBuilder from MCP prompt result", async () => {
  const context = new AIGNE().newContext();

  const prompt: GetPromptResult = {
    description: "Test prompt",
    messages: [
      {
        role: "user",
        content: {
          type: "text",
          text: "Hello",
        },
      },
      {
        role: "user",
        content: {
          type: "resource",
          resource: {
            uri: "echo://Hello",
            text: "Resource echo: Hello",
            mimeType: "text/plain",
          },
        },
      },
      {
        role: "user",
        content: {
          type: "resource",
          resource: {
            uri: "test://image-resource",
            blob: Buffer.from("FAKE IMAGE FROM RESOURCE").toString("base64"),
            mimeType: "image/png",
          },
        },
      },
      {
        role: "user",
        content: {
          type: "image",
          mimeType: "image/png",
          data: Buffer.from("FAKE IMAGE").toString("base64"),
        },
      },
      {
        role: "assistant",
        content: {
          type: "text",
          text: "How can I help you?",
        },
      },
    ],
  };

  const promptBuilder = PromptBuilder.from(prompt);
  expect(await promptBuilder.build({ context })).toEqual(
    expect.objectContaining({
      messages: [
        {
          role: "user",
          content: "Hello",
        },
        {
          role: "user",
          content: "Resource echo: Hello",
        },
        {
          role: "user",
          content: [
            { type: "image_url", url: Buffer.from("FAKE IMAGE FROM RESOURCE").toString("base64") },
          ],
        },
        {
          role: "user",
          content: [{ type: "image_url", url: Buffer.from("FAKE IMAGE").toString("base64") }],
        },
        {
          role: "agent",
          content: "How can I help you?",
        },
      ],
    }),
  );
});

test("PromptBuilder from file", async () => {
  const context = new AIGNE().newContext();

  const path = join(import.meta.dirname, "test-prompt.txt");
  const content = await readFile(path, "utf-8");

  const builder = PromptBuilder.from({ path });

  const prompt = await builder.build({ input: { agentName: "Alice" }, context });

  expect(prompt).toEqual({
    messages: [
      {
        role: "system",
        content: content.replace("{{agentName}}", "Alice"),
      },
    ],
  });
});
