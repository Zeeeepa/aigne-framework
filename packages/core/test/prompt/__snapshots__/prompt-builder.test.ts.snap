// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`PromptBuilder from file 1`] = `
{
  "messages": [
    {
      "content": 
"## instructions

You are Alice, a professional customer service.

Output in English

"
,
      "name": undefined,
      "role": "system",
    },
  ],
  "modelOptions": undefined,
  "responseFormat": undefined,
  "toolAgents": undefined,
  "toolChoice": undefined,
  "tools": undefined,
}
`;
